### PR TITLE
Start/stop watching individual paths

### PR DESCRIPTION
**Pull request by lptr**

### CODE REVIEW

**Comment by lptr** [comment]
@wolfs I'll merge this now, but please take a look when you are back.

File: src/file-events/headers/apple_fsnotifier.h [hunk 1]
```diff
@@ -32,10 +32,13 @@ class WatchPoint {
...
 public:
-    Server(JNIEnv* env, jobject watcherCallback);
+    Server(JNIEnv* env, jobject watcherCallback, long latencyInMillis);
     ~Server();
 
-    void startWatching(const u16string& path, long latencyInMillis);
+    void startWatching(const u16string& path);
+    void stopWatching(const u16string& path);
+
+    // TODO This should be private
```
Comment (wolfs) [inline src/file-events/headers/apple_fsnotifier.h]: Can we make it private? And maybe use `friend` for the static method which calls it?

File: src/main/java/net/rubygrapefruit/platform/internal/jni/OsxFileEventFunctions.java [hunk 2]
```diff
@@ -60,16 +59,11 @@
...
      */
     // TODO How to set kFSEventStreamCreateFlagNoDefer when latency is non-zero?
-    public FileWatcher startWatching(Collection<String> paths, final long latency, final TimeUnit unit, FileWatcherCallback callback) {
-        return createWatcher(paths, callback, new WatcherFactory() {
-            @Override
-            public FileWatcher createWatcher(String[] canonicalPaths, NativeFileWatcherCallback callback) {
-                return startWatching(canonicalPaths, unit.toMillis(latency), callback);
-            }
-        });
+    public FileWatcher startWatcher(long latency, TimeUnit unit, FileWatcherCallback callback) {
```
Comment (wolfs) [inline src/main/java/net/rubygrapefruit/platform/internal/jni/OsxFileEventFunctions.java]: Should this be called `startWatcher`? We start the thread there, though no watching actually starts.

File: src/file-events/cpp/win_fsnotifier.cpp [hunk 3]
```diff
@@ -8,14 +10,25 @@ using namespace std;
...
     this->buffer = (FILE_NOTIFY_INFORMATION*) malloc(EVENT_BUFFER_SIZE);
     ZeroMemory(&this->overlapped, sizeof(OVERLAPPED));
     this->overlapped.hEvent = this;
     this->directoryHandle = directoryHandle;
     this->status = WATCH_UNINITIALIZED;
+    unique_lock<mutex> lock(listenerMutex);
+    QueueUserAPC(startWatchCallback, serverThreadHandle, (ULONG_PTR) this);
+    listenerStarted.wait(lock);
+    if (status != WATCH_LISTENING) {
+        throw new FileWatcherException("Couldn't start listening");
```
Comment (wolfs) [inline src/file-events/cpp/win_fsnotifier.cpp]: Can we add the path we try to watch to this exception?

File: src/file-events/cpp/win_fsnotifier.cpp [hunk 4]
```diff
@@ -182,25 +183,18 @@ void Server::startWatching(JNIEnv* env, const u16string& path) {
...
-        case WATCH_LISTENING:
-            watchPoints.push_back(watchPoint);
-            break;
-        default:
-            log_severe(env, "Couldn't start listening to '%ls': %d", pathW.c_str(), ret);
-            delete watchPoint;
-            // TODO Error handling
-            break;
-    }
+    watchPoints.emplace(piecewise_construct,
```
Comment (wolfs) [inline src/file-events/cpp/win_fsnotifier.cpp]: Should we also check here if the location is already watched?

File: src/file-events/cpp/win_fsnotifier.cpp [hunk 5]
```diff
@@ -182,25 +183,18 @@ void Server::startWatching(JNIEnv* env, const u16string& path) {
...
+        forward_as_tuple(path),
+        forward_as_tuple(this, path, directoryHandle, threadHandle));
 }
 
-void Server::reportFinished(WatchPoint* watchPoint) {
-    watchPoints.remove(watchPoint);
-    delete watchPoint;
+void Server::stopWatching(JNIEnv* env, const u16string& path) {
+    watchPoints.find(path)->second.close();
```
Comment (wolfs) [inline src/file-events/cpp/win_fsnotifier.cpp]: Should we check here if we are watching the path?

File: src/test/groovy/net/rubygrapefruit/platform/file/AbstractFileEventsTest.groovy [hunk 6]
```diff
@@ -280,6 +280,21 @@ abstract class AbstractFileEventsTest extends Specification {
         expectedChanges.await()
     }
 
+    def "does not receive events after directory is unwatched"() {
```
Comment (wolfs) [inline src/test/groovy/net/rubygrapefruit/platform/file/AbstractFileEventsTest.groovy]: There are tests missing for what happens when
- `startWatching` twice for the same path
- `stopWatching` for a non-watched path
- `stopWatching` twice for the same watched path.

File: src/file-events/cpp/apple_fsnotifier.cpp [hunk 7]
```diff
@@ -190,29 +191,27 @@ void Server::handleEvent(JNIEnv* env, char* path, FSEventStreamEventFlags flags)
     reportChange(env, type, pathStr);
 }
 
-void Server::startWatching(const u16string& path, long latencyInMillis) {
-    watchPoints.emplace_back(this, threadLoop, path, latencyInMillis);
+void Server::startWatching(const u16string& path) {
+    if (watchPoints.find(path) != watchPoints.end()) {
+        throw new FileWatcherException("Already watching path");
+    }
+    // TODO Is this necessary
```
[the remark this answers is not in the document]
Comment (lptr) [reply src/file-events/cpp/apple_fsnotifier.cpp]: I just checked with Thad, and it seems to be. I'm removing the TODOs soon.

File: src/file-events/cpp/apple_fsnotifier.cpp [hunk 8]
```diff
@@ -190,29 +191,27 @@ void Server::handleEvent(JNIEnv* env, char* path, FSEventStreamEventFlags flags)
     reportChange(env, type, pathStr);
 }
 
-void Server::startWatching(const u16string& path, long latencyInMillis) {
-    watchPoints.emplace_back(this, threadLoop, path, latencyInMillis);
+void Server::startWatching(const u16string& path) {
+    if (watchPoints.find(path) != watchPoints.end()) {
+        throw new FileWatcherException("Already watching path");
```
[the remark this answers is not in the document]
Comment (lptr) [reply src/file-events/cpp/apple_fsnotifier.cpp]: I think we should opt for strictness in the beginning. If it becomes troublesome we can make things more lenient, but I'm not sure what would convince me that we need to do just that.

The Java side should be aware of all the watchers that are registered, and it should signal a problem if we tried to register the same watcher twice, so I think it'll be fine to fail always.

[the remark this answers is not in the document]
Comment (lptr) [reply src/file-events/cpp/apple_fsnotifier.cpp]: I made a not to add Javadocs.

File: src/file-events/headers/apple_fsnotifier.h (hunk 1 above)
Comment (lptr) [reply src/file-events/headers/apple_fsnotifier.h]: We will, yeah.

File: src/main/java/net/rubygrapefruit/platform/internal/jni/OsxFileEventFunctions.java (hunk 2 above)
Comment (lptr) [reply src/main/java/net/rubygrapefruit/platform/internal/jni/OsxFileEventFunctions.java]: That's a good question. It's a bit painful to rename things around here, so I didn't play around with it yet, but I definitely want to clarify the terminology once we have Linux done as well.

File: src/file-events/cpp/win_fsnotifier.cpp (hunk 3 above)
Comment (lptr) [reply src/file-events/cpp/win_fsnotifier.cpp]: No. Dynamic strings are not a thing in C++ exceptions unfortunately. :/

Comment (wolfs) [reply src/file-events/cpp/win_fsnotifier.cpp]: :(

File: src/file-events/cpp/win_fsnotifier.cpp (hunk 4 above)
Comment (lptr) [reply src/file-events/cpp/win_fsnotifier.cpp]: Good point. Fixing it.

File: src/file-events/cpp/win_fsnotifier.cpp (hunk 5 above)
Comment (lptr) [reply src/file-events/cpp/win_fsnotifier.cpp]: Yes, fixing.

File: src/test/groovy/net/rubygrapefruit/platform/file/AbstractFileEventsTest.groovy (hunk 6 above)
Comment (lptr) [reply src/test/groovy/net/rubygrapefruit/platform/file/AbstractFileEventsTest.groovy]: Good point, adding.

**Pull request merged**